### ISSUE TITLE
coordinator stopped while threads still running

### ISSUE THREAD

**Issue by dleebrown** [dleebrown/ANNA#50]
Never seen this before. It happens in the xval subloop.

**Comment by dleebrown** [comment]
Traceback (most recent call last):
  File "/data/donald/PYTHON/ANNA/current/THESIS_ANNA/ANNA_train.py", line 559, in <module>
    train_neural_network(parameters)
  File "/data/donald/PYTHON/ANNA/current/THESIS_ANNA/ANNA_train.py", line 531, in train_neural_network
    float(parameters['KEEP_PROB1']), bsize_train1, inherit_iter_count=0)
  File "/data/donald/PYTHON/ANNA/current/THESIS_ANNA/ANNA_train.py", line 467, in train_loop
    xvcost = xval_subloop(learn_rate, xv_size, i, inherit_iter_count)
  File "/data/donald/PYTHON/ANNA/current/THESIS_ANNA/ANNA_train.py", line 405, in xval_subloop
    xvcoordinator.join(xval_threads)
  File "/home/donald/anaconda2/lib/python2.7/site-packages/tensorflow/python/training/coordinator.py", line 397, in join
    " ".join(stragglers))
RuntimeError: Coordinator stopped with threads still running: Thread-32


the coordinator is requested to stop before the join command, so not sure what is up here. will try to replicate. using xval batch size of 3000.

**Comment by dleebrown** [comment]
trying to get a handle on this bug. it can only be reproduced for sufficiently large cross-validation sample sizes. so I think what is happening is the cross-validation coordinator is joining the threads before they are finished, causing a deadlock. Not sure how to fix, but this thread may be useful:

https://github.com/tensorflow/tensorflow/issues/2130

ANNA is humming along with 500 for the xval size.

**Comment by dleebrown** [comment]
Issue does occur with smaller batch sizes, just less frequently.

**Comment by dleebrown** [comment]
So it looks like it is a deadlock issue, where the xval enqueue operation tries to fill the xval queue, but it is already full. This causes the xval thread to hang and the tensorflow coordinator can't join the xval threads. 

I have a workaround for now: I set a timeout on the session.run(xval queue op) and caught the error with an exception that just runs xval.queue(close pending enqueues=true). This forces the xval queue to close, which allows the xval thread to join and the training process to proceed. If this workaround seems stable I'll close this issue.

**Comment by dleebrown** [comment]
Solved the issue. Confirmed that what was happening was the xval enqueue thread would block when the xval queue was full. Using an exception to close the queue would close the queue permanently and the code would run until the xval queue was empty, and then throw an error. 

Solution: I added a conditional before session.run xval enqueue call that only executed the enqueue op if the queue had enough space to enqueue an xval batch. Thus, when the xval queue is full, the code just skips the enqueue op and avoids the block.
